virtcontainers/vm.go

```diff
@@ -0,0 +1,233 @@
+}
+
+// VMConfig is a collection of all info that a new blackbox VM needs.
+type VMConfig struct {
+	HypervisorType   HypervisorType
+	HypervisorConfig HypervisorConfig
+
+	AgentType   AgentType
+	AgentConfig interface{}
+}
+
+// Valid check VMConfig validity.
+func (c *VMConfig) Valid() error {
+	return c.HypervisorConfig.valid()
+}
+
+// NewVM creates a new VM based on provided VMConfig.
+func NewVM(config VMConfig) (*VM, error) {
+	hypervisor, err := newHypervisor(config.HypervisorType)
+	if err != nil {
+		return nil, err
+	}
+
+	err = config.Valid()
+	if err != nil {
+		return nil, err
+	}
+
+	id := uuid.Generate().String()
+
+	virtLog.WithField("vm id", id).WithField("config", config).Info("create new vm")
+
+	defer func() {
+		if err != nil {
+			virtLog.WithField("vm id", id).WithError(err).Error("failed to create new vm")
+		}
+	}()
+
+	err = hypervisor.init(id, &config.HypervisorConfig, Resources{}, &filesystem{})
+	if err != nil {
+		return nil, err
+	}
+
+	err = hypervisor.createSandbox()
+	if err != nil {
+		return nil, err
+	}
+
+	agent := newAgent(config.AgentType)
+	agentConfig := newAgentConfig(config.AgentType, config.AgentConfig)
+	// do not keep connection for temp agent
+	if c, ok := agentConfig.(KataAgentConfig); ok {
+		c.LongLiveConn = false
+	}
+	vmSharePath := buildVMSharePath(id)
+	err = agent.configure(hypervisor, id, vmSharePath, true, agentConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	err = hypervisor.startSandbox()
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err != nil {
+			virtLog.WithField("vm id", id).WithError(err).Info("clean up vm")
+			hypervisor.stopSandbox()
+		}
+	}()
+
+	// VMs booted from template are paused, do not check
+	if !config.HypervisorConfig.BootFromTemplate {
+		err = hypervisor.waitSandbox(vmStartTimeout)
+		if err != nil {
+			return nil, err
+		}
+
+		virtLog.WithField("vm id", id).Info("check agent status")
+		err = agent.check()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &VM{
+		id:         id,
+		hypervisor: hypervisor,
+		agent:      agent,
+		cpu:        config.HypervisorConfig.DefaultVCPUs,
+		memory:     config.HypervisorConfig.DefaultMemSz,
+	}, nil
+}
+
+func buildVMSharePath(id string) string {
+	return filepath.Join(RunVMStoragePath, id, "shared")
+}
+
+func (v *VM) logger() logrus.FieldLogger {
+	return virtLog.WithField("vm id", v.id)
+}
+
+// Pause pauses a VM.
+func (v *VM) Pause() error {
+	v.logger().Info("pause vm")
+	return v.hypervisor.pauseSandbox()
+}
+
+// Save saves a VM to persistent disk.
+func (v *VM) Save() error {
+	v.logger().Info("save vm")
+	return v.hypervisor.saveSandbox()
+}
+
+// Resume resumes a paused VM.
+func (v *VM) Resume() error {
+	v.logger().Info("resume vm")
+	return v.hypervisor.resumeSandbox()
+}
+
+// Start kicks off a configured VM.
+func (v *VM) Start() error {
+	v.logger().Info("start vm")
+	return v.hypervisor.startSandbox()
+}
+
+// Kill stops a VM process.
+func (v *VM) Kill() error {
```

Comment:
I would prefer `Stop()` than `Kill()` here since I think more about the VM semantic being start/stop. The fact the VM is a process should be a detail here, hence no reason to call it `Kill()`